### PR TITLE
Refactor debug-warning rules

### PR DESCRIPTION
**Pull request by LecrisUT**
- Separated the rules to be independent of each other and made advanced variables to control their settings
- Made the macro definition `PRIVATE` because they only affect the library compilation and not `spglib.h`
- Simplified various debug statements since most of them are optimized by compiler when empty. Good coding practices these days is to avoid `#ifdef` and macros as much as possible.
- @atztogo you have copyright of `niggli.c` and it said to be part of a different project, so I have to ask your approval for 32bae2be15f9161d060b0105316f1b9b052d7cee. Generally I avoid including code from project to project and then editing it, and when I do copy around and have full copyright, I tend to remove and migrate the copyright to avoid such confusions
- What's the deal with c2efc49923cb8d70f54ef53d7274d8afe7dd89db? Those functions are not used anywhere. Is it for future implementation, in which case it should be moved to a separate file, and be marked for experimental builds

### CODE REVIEW

**Comment by atztogo** [comment]
> * @atztogo you have copyright of `niggli.c` and it said to be part of a different project, so I have to ask your approval for [32bae2b]. Generally I avoid including code from project to project and then editing it, and when I do copy around and have full copyright, I tend to remove and migrate the copyright to avoid such confusions

It's perfectly no problem. My aim was to make it a stand alone routine. I should do like you wrote. We can migrate the copyright.

**Comment by LecrisUT** [comment]
Thanks for the clarification. I should leave that commit to you. You can simply pull my branch and push directly to it.

**Comment by atztogo** [comment]
> What's the deal with https://github.com/spglib/spglib/commit/c2efc49923cb8d70f54ef53d7274d8afe7dd89db? Those functions are not used anywhere. Is it for future implementation, in which case it should be moved to a separate file, and be marked for experimental builds

If it is who I wrote, I think it can be removed.

**Comment by atztogo** [comment]
> Thanks for the clarification. I should leave that commit to you. You can simply pull my branch and push directly to it.

Sorry for bothering you. How can I do it concretely?

**Comment by LecrisUT** [comment]
> > What's the deal with [c2efc49](https://github.com/spglib/spglib/commit/c2efc49923cb8d70f54ef53d7274d8afe7dd89db)? Those functions are not used anywhere. Is it for future implementation, in which case it should be moved to a separate file, and be marked for experimental builds
> 
> If it is who I wrote, I think it can be removed.

It goes all the way back to "First commit on github" 10 year ago:

> Sorry for bothering you. How can I do it concretely?

No problem at all:
```console
Assuming you push using ssh, otherwise change it to https
$ git remote add LecrisUT git@github.com:LecrisUT/spglib.git
$ git fetch LecrisUT
$ git checkout LecrisUT/debug-warnings
Edit to your heart's content
$ vi src/niggli.c
Then do the usual
$ git commit -s
Now just specify my remote instead of origin
$ git push LecrisUT
```

**Comment by atztogo** [comment]
> It goes all the way back to "First commit on github" 10 year ago:

Then, it's definitely me. Let's remove it. Spglib was born in Aachen more than 10 years ago.

Thanks for your instruction about git.

File: src/primitive.c
Comment (atztogo) [inline src/primitive.c:231]: Do you assume running this loop is always negligible in computing time? Probably yes, but if there is no problem, we can keep `#ifdef SPGDEBUG`.

Comment (atztogo) [reply src/primitive.c:231]: I re-read your initial PR comments. Probably your advice `Good coding practices these days is to avoid #ifdef` can be applied here, too.

**Review by atztogo — APPROVED** [review]

**Comment by atztogo** [comment]
@LecrisUT, thanks for all of these improvements.

File: src/primitive.c
Comment (lan496) [reply src/primitive.c:231]: I am just curious when these redundant lines are promised to be omitted, -O2 and -O3? Is there a reference for that?

Comment (LecrisUT) [reply src/primitive.c:231]: I think this can be tested out on godlbolt.org. Let me cook up an example file

Comment (LecrisUT) [reply src/primitive.c:231]: Ok Here we are: https://godbolt.org/z/xqE9rhq5b
So as little as `-O1` (See bottom 2 panels where left is completely optimized, while right the loop is running)

Comment (LecrisUT) [reply src/primitive.c:231]: > I re-read your initial PR comments. Probably your advice "Good coding practices these days is to avoid `#ifdef`" can be applied here, too.

Indeed, I very much tried to think of a way around it, but I couldn't find a way to make the printing conditional. I did consider getting rid of those variable and using `NDEBUG`, but it didn't seem compatible with how `warning_print` is used. That's why I've made them into separate cmake options

**Comment by LecrisUT** [comment]
Actually I was looking at the debug in the tests and I find the debug messages too spamy to be useful to navigate so I want to get your opinion about it (not for this PR). Have you used the breakpoint debugger in VSCode? Do you find it more useful than the manual printing? I find that the usage of all of the `debug_print` and `warning_print` are equivalent with what you achieve with breakpoint debugger, and if these can be eliminated in favor of the latter, it will make debugging of the test cases much more manageable.

**Comment by atztogo** [comment]
Of course I am happy to move to the modern world!

**Comment by atztogo** [comment]
When I started to learn C, google didn't exist.

**Comment by LecrisUT** [comment]
Great, just wanted to make sure if it's not part of a developer's active setup, and if it is, if they need some help setting up the IDE debugger to work around it. I will leave it to @lan496 to merge this one after he's caught up with the discussion.

File: src/primitive.c
Comment (lan496) [reply src/primitive.c:231]: Thank you for your detailed explanation. It's interesting that even O1 completely optimizes the code.

**Review by lan496 — APPROVED** [review]
LGTM!

**Pull request merged**
